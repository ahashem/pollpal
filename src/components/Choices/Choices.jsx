import React from 'react';
import PropTypes from 'prop-types';
import { List, Progress, Radio, Skeleton } from 'antd';
import Helpers from '../../utils/helpers';

const { Item } = List;
const RadioButton = Radio.Button;
const RadioGroup = Radio.Group;


const Choices = ({ onChange, loading, choices }) => {
  return (
    <RadioGroup
      onChange={onChange}
    >
      <List
        itemLayout="horizontal"
        loading={loading}
        dataSource={choices}
        renderItem={choice => (
          <RadioButton
            value={choice.url}
            buttonStyle="solid"
          >
            <Item key={choice.url}>
              <Skeleton loading={loading} active>
                <Item.Meta
                  title={choice.choice}
                  description={`${choice.votes} votes`}
                />
                <div style={{ width: 120, marginRight: 20 }}>
                  <Progress
                    percent={Helpers.calculatePercentage(choice.votes, Helpers.reduceSum(choices, 'votes'))}
                    size="small"
                    status="active"
                    default="small"
                  />
                </div>
              </Skeleton>
            </Item>
          </RadioButton>
        )}
      />
    </RadioGroup>
  )
    ;
};

Choices.propTypes = {
  onChange: PropTypes.func,
  loading: PropTypes.bool,
  choices: PropTypes.arrayOf(PropTypes.shape({
    choice: PropTypes.string,
    url: PropTypes.string,
    votes: PropTypes.number,
  })).isRequired,
};

export default Choices;
