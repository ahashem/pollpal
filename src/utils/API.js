import 'whatwg-fetch';
import { CREATED, OK } from 'http-status';

/**
 *
 * @param url
 * @param method
 * @param data
 * @return {Promise<any>}
 */
const realAPIRequest = (url, method = 'GET', data = {}) => {
  return new Promise((resolve, reject) => {

    const options = {
      method,
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        'Access-Control-Allow-Methods': 'GET,POST,HEAD,OPTIONS,PUT,DELETE',
      },
    };
    // If there is data passed assign it to the request body
    if (Object.keys(data).length > 0) options.body = data;

    // Add token to every request if there's at localStorage
    fetch(url, options)
      .then((response) => {

        // Format response
        return {
          status: response.status,
          text: response.statusText,
          url: response.url,
          data: response.json(),
        };

      })
      .then((response) => {
        switch (response.status) {
          case OK:
          case CREATED:
            resolve(response.data);
            break;
          case /^4[0-9][0-9]$/:
          case /^5[0-9][0-9]$/:
            return response.data
              .then((res) => {
                const error = {
                  errorCode: response.status,
                  errorMsg: res.error || res.message,
                };
                throw new Error(JSON.stringify(error));
              })
              .catch(error => reject(error));
          default:
            resolve(response.data);
            break;
        }
      })
      .catch(error => reject(error));
  });
};

const mockRequest = (mockDataPath) => {
  return new Promise((resolve, reject) => resolve(mockDataPath));
};

const request = (url, method = 'GET', data = {}) => {
  return (process.env.REACT_APP_MOCK_API) ? mockRequest(url) : realAPIRequest(url, method = 'GET', data = {});
};

const basePollApiURL = `${process.env.REACT_APP_POLLS_BASE_URL}`;
const questionsApiURL = `${basePollApiURL}/questions`;
const questionsApi = {
  getAllQuestions() {
    return request(questionsApiURL, 'GET');
  },
  getQuestionById(id) {
    return request(`${questionsApiURL}/${id}`, 'GET');
  },
  voteOnChoice(voteUrl){
    return request(`${basePollApiURL}${voteUrl}`, "POST");
  },
};


export { request as default, questionsApi };
