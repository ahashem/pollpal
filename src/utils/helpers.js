const Helpers = {

  /**
   *
   * @param {Number|string} number
   * @param {Number|string} total
   * @return {Number|string}
   */
  calculatePercentage(number, total) {
    if (total !== 0) {
      return ((number / total) * 100).toFixed(2);
    }
    return 0;
  },

  /**
   *
   * @param {string|Date} date
   * @return {string}
   */
  formatDate(date) {
    const newDate = new Date(date);
    if (newDate !== 'Invalid Date') {
      return newDate.toLocaleString('en-us', { day: 'numeric', month: 'short', year: 'numeric' });
    }
    return date;
  },

  /**
   * Simple Array sum reducer that returns the Sum of all `key` values in Array elements
   * @param {Array} array
   * @param {string} key
   * @return {Number|string}
   */
  reduceSum(array, key) {
    return array.reduce((acc, curr) => acc + curr[`${key}`], 0);
  }
};

export default Helpers
