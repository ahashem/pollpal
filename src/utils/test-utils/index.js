import mockQuestions from './mock-test-quesions';

export { mockQuestions };
