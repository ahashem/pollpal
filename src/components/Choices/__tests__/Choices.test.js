import React from 'react';
import { shallow } from 'enzyme/build';

import Choices from '../Choices';

import { mockQuestions } from '../../../utils/test-utils';

describe('<Choices /> Component', () => {
  it('should render children and match snapshot', () => {
    const mockOnChange = jest.fn();

    const rendered = shallow(<Choices
      choices={mockQuestions[0].choices}
      onChange={mockOnChange}
      loading={false}
    />);

    expect(rendered).toMatchSnapshot();
  });
});
