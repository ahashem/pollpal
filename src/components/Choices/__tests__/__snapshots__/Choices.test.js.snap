// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Choices /> Component should render children and match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Choices
    choices={
      Array [
        Object {
          "choice": "Swift",
          "url": "/questions/1/choices/1",
          "votes": 2048,
        },
        Object {
          "choice": "Python",
          "url": "/questions/1/choices/2",
          "votes": 1024,
        },
        Object {
          "choice": "Objective-C",
          "url": "/questions/1/choices/3",
          "votes": 512,
        },
        Object {
          "choice": "Ruby",
          "url": "/questions/1/choices/4",
          "votes": 256,
        },
      ]
    }
    loading={false}
    onChange={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttonStyle": "outline",
      "children": <List
        bordered={false}
        dataSource={
          Array [
            Object {
              "choice": "Swift",
              "url": "/questions/1/choices/1",
              "votes": 2048,
            },
            Object {
              "choice": "Python",
              "url": "/questions/1/choices/2",
              "votes": 1024,
            },
            Object {
              "choice": "Objective-C",
              "url": "/questions/1/choices/3",
              "votes": 512,
            },
            Object {
              "choice": "Ruby",
              "url": "/questions/1/choices/4",
              "votes": 256,
            },
          ]
        }
        itemLayout="horizontal"
        loading={false}
        pagination={false}
        prefixCls="ant-list"
        renderItem={[Function]}
        split={true}
      />,
      "disabled": false,
      "onChange": [MockFunction],
      "prefixCls": "ant-radio",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bordered": false,
        "dataSource": Array [
          Object {
            "choice": "Swift",
            "url": "/questions/1/choices/1",
            "votes": 2048,
          },
          Object {
            "choice": "Python",
            "url": "/questions/1/choices/2",
            "votes": 1024,
          },
          Object {
            "choice": "Objective-C",
            "url": "/questions/1/choices/3",
            "votes": 512,
          },
          Object {
            "choice": "Ruby",
            "url": "/questions/1/choices/4",
            "votes": 256,
          },
        ],
        "itemLayout": "horizontal",
        "loading": false,
        "pagination": false,
        "prefixCls": "ant-list",
        "renderItem": [Function],
        "split": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttonStyle": "outline",
        "children": <List
          bordered={false}
          dataSource={
            Array [
              Object {
                "choice": "Swift",
                "url": "/questions/1/choices/1",
                "votes": 2048,
              },
              Object {
                "choice": "Python",
                "url": "/questions/1/choices/2",
                "votes": 1024,
              },
              Object {
                "choice": "Objective-C",
                "url": "/questions/1/choices/3",
                "votes": 512,
              },
              Object {
                "choice": "Ruby",
                "url": "/questions/1/choices/4",
                "votes": 256,
              },
            ]
          }
          itemLayout="horizontal"
          loading={false}
          pagination={false}
          prefixCls="ant-list"
          renderItem={[Function]}
          split={true}
        />,
        "disabled": false,
        "onChange": [MockFunction],
        "prefixCls": "ant-radio",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bordered": false,
          "dataSource": Array [
            Object {
              "choice": "Swift",
              "url": "/questions/1/choices/1",
              "votes": 2048,
            },
            Object {
              "choice": "Python",
              "url": "/questions/1/choices/2",
              "votes": 1024,
            },
            Object {
              "choice": "Objective-C",
              "url": "/questions/1/choices/3",
              "votes": 512,
            },
            Object {
              "choice": "Ruby",
              "url": "/questions/1/choices/4",
              "votes": 256,
            },
          ],
          "itemLayout": "horizontal",
          "loading": false,
          "pagination": false,
          "prefixCls": "ant-list",
          "renderItem": [Function],
          "split": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
